pkg/operator/controller/extension/virtualcluster/reconciler.go

```diff
@@ -0,0 +1,316 @@
+	err := r.reconcileVirtualClusterResources(ctx, log, gardenClient, extension)
+	if err != nil {
+		conditions.virtualClusterReconciled = v1beta1helper.UpdatedConditionWithClock(r.Clock, conditions.virtualClusterReconciled, gardencorev1beta1.ConditionFalse, ConditionReconcileFailed, err.Error())
+	} else {
+		conditions.virtualClusterReconciled = v1beta1helper.UpdatedConditionWithClock(r.Clock, conditions.virtualClusterReconciled, gardencorev1beta1.ConditionTrue, ConditionReconcileSuccess, fmt.Sprintf("Extension %q has been reconciled successfully", extension.Name))
+	}
+
+	return errors.Join(err, r.updateExtensionStatus(ctx, log, extension, conditions))
+}
+
+func (r *Reconciler) reconcileVirtualClusterResources(ctx context.Context, log logr.Logger, gardenClient client.Client, extension *operatorv1alpha1.Extension) error {
+	// return early if we do not have to make a deployment
+	if extension.Spec.Deployment == nil ||
+		extension.Spec.Deployment.ExtensionDeployment == nil ||
+		extension.Spec.Deployment.ExtensionDeployment.Helm == nil {
+		return r.deleteVirtualClusterResources(ctx, log, gardenClient, extension)
+	}
+
+	if err := r.reconcileControllerDeployment(ctx, gardenClient, extension); err != nil {
+		err := fmt.Errorf("failed to reconciler controller deployment: %w", err)
+		return err
+	}
+	r.Recorder.Event(extension, corev1.EventTypeNormal, "Reconciliation", "ControllerDeployment applied successfully")
+
+	if err := r.reconcileControllerRegistration(ctx, gardenClient, extension); err != nil {
+		err := fmt.Errorf("failed to reconciler controller registration: %w", err)
+		return err
+	}
+	r.Recorder.Event(extension, corev1.EventTypeNormal, "Reconciliation", "ControllerRegistration applied successfully")
+	return nil
+}
+
+func (r *Reconciler) reconcileControllerDeployment(ctx context.Context, gardenClient client.Client, extension *operatorv1alpha1.Extension) error {
+	ctrlDeploy := &gardencorev1.ControllerDeployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: extension.Name,
+		},
+	}
+
+	deployMutateFn := func() error {
+		ctrlDeploy.Helm = &gardencorev1.HelmControllerDeployment{
+			Values:        extension.Spec.Deployment.ExtensionDeployment.Values,
+			OCIRepository: extension.Spec.Deployment.ExtensionDeployment.Helm.OCIRepository,
+		}
+		return nil
+	}
+
+	if _, err := controllerutil.CreateOrUpdate(ctx, gardenClient, ctrlDeploy, deployMutateFn); err != nil {
+		return fmt.Errorf("failed to create or update ControllerDeployment: %w", err)
+	}
+	return nil
+}
+
+func (r *Reconciler) reconcileControllerRegistration(ctx context.Context, gardenClient client.Client, extension *operatorv1alpha1.Extension) error {
+	ctrlReg := &gardencorev1beta1.ControllerRegistration{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: extension.Name,
+		},
+	}
+
+	regMutateFn := func() error {
+		// handle well known annotations
+		if v, ok := extension.Annotations[v1beta1constants.AnnotationPodSecurityEnforce]; ok {
```

Comment:
I don't understand this in the light of https://github.com/gardener/gardener/pull/10006/files#diff-c68ac578492b3e3ae848fde54c007925a0b8a726c930d26a5926aafc9b4a8fb1R29